File: Olympiad/COCI/official/2007/contest3_solutions/lista.pas

/*
  Croatian Open Competition in Informatics 2006/2007
  Contest 3 - Task LISTA
  Programming language Pascal
*/

program lista;

type cvor = record
   L, R : longint;
end;

var
   n, k : longint;
   list : array[0..500001] of cvor;

   prosli : array[0..500000] of longint;
   M : array[0..500000] of longint;

   A : array[0..500000] of longint;

   c : char;
   x, y : longint;
   i, j : longint;
   lo, hi, mid : longint;
   najdulji : longint;
   zadnji : longint;

begin
   readln( n, k );
   for i := 0 to n+1 do begin
      list[i].L := i-1;
      list[i].R := i+1;
   end;

   for i := 1 to k do begin
      readln( c, x, y );

      { izbaci x iz liste }
      list[list[x].R].L := list[x].L;
      list[list[x].L].R := list[x].R;

      { odredi nove susjede cvoru x }
      if c = 'A' then begin
         list[x].L := list[y].L;
         list[x].R := y;
      end else begin
         list[x].L := y;
         list[x].R := list[y].R;
      end;

      { ubaci x u listu }
      list[list[x].R].L := x;
      list[list[x].L].R := x;
   end;

   { nadji najdulji rastuci podniz }
   M[0] := 0;
   for i := 1 to n do M[i] := 1000000000;

   najdulji := 0;
   x := list[0].R;
   while x <> n+1 do begin
      lo := 0; hi := n;
      while lo <> hi do begin
         mid := (lo+hi) div 2;
         if M[mid] < x then lo := mid + 1 else hi := mid;
      end;
      M[lo] := x;
      prosli[x] := M[lo-1];
      if lo > najdulji then najdulji := lo;

      x := list[x].R;
   end;

   writeln( n - najdulji );

   k := 0;
   x := M[najdulji];
   while x <> 0 do begin
      A[najdulji-k] := x;
      k := k + 1;

      x := prosli[x];
   end;

   i := 1; j := 1;
   while i <= najdulji do begin
      if j = A[i] then i := i + 1
      else writeln( 'A ', j, ' ', A[i] );
      j := j + 1;
   end;

   zadnji := A[najdulji];
   while j <= n do begin
      writeln( 'B ', j, ' ', zadnji );
      zadnji := j;
      j := j + 1;
   end;
end.

File: Olympiad/COCI/official/2007/contest3_solutions/lista.cpp

/*
  Croatian Open Competition in Informatics 2006/2007
  Contest 3 - Task LISTA
  Programming language C++
*/

#include <algorithm>
#include <cstdio>
#include <vector>

using namespace std;

#define MAX 500000

struct cvor { int L, R; } lista[MAX+2];

int prosli[MAX+1];
int M[MAX+1];

int main( void ) {
   int n, m;
   scanf( "%d%d", &n, &m );
   for( int i = 0; i <= n+1; ++i ) {
      lista[i].L = i-1;
      lista[i].R = i+1;
   }

   for( int i = 0; i < m; ++i ) {
      char c; int x, y;
      scanf( " %c%d%d", &c, &x, &y );

      // izbaci x iz liste
      lista[lista[x].R].L = lista[x].L;
      lista[lista[x].L].R = lista[x].R;

      // odredi nove susjede cvoru x
      if( c == 'A' ) {
         lista[x].L = lista[y].L;
         lista[x].R = y;
      } else {
         lista[x].L = y;
         lista[x].R = lista[y].R;
      }

      // ubaci x u listu
      lista[lista[x].R].L = x;
      lista[lista[x].L].R = x;
   }

   // nadji najdulji rastuci podniz
   M[0] = 0;
   for( int i = 1; i <= n; ++i ) M[i] = 1000000000;

   int najdulji = 0;
   for( int x = lista[0].R; x != n+1; x = lista[x].R ) {

      int d = upper_bound( M, M+n, x ) - M;
      M[d] = x;
      prosli[x] = M[d-1];
      najdulji >?= d;
   }

   printf( "%d\n", n - najdulji );

   vector<int> A;
   for( int x = M[najdulji]; x != 0; x = prosli[x] )
      A.push_back( x );
   reverse( A.begin(), A.end() );

   // ispisi niz operacija
   int i = 0, j = 1;
   while( i < najdulji ) {
      if( j == A[i] ) ++i;
      else printf( "A %d %d\n", j, A[i] );
      ++j;
   }
   int zadnji = A.back();
   while( j <= n ) {
      printf( "B %d %d\n", j, zadnji );
      zadnji = j;
      ++j;
   }

   return 0;
}
